### PR TITLE
upgrade ruff

### PR DESCRIPTION
**Pull request by chassing**
- **bump ruff**
- **remove obsolete reconcile.utils.differ**
- **:rotating_light: fix ruff errors**

## Summary by CodeRabbit

* **Chores**
  * Upgraded linting tooling across projects and consolidated diff utilities into a shared implementation; adjusted lint ignore rules.
  * Removed legacy differ tests and updated project lint config.

* **Refactor**
  * Modernized type annotations and context-manager typings; standardized early-return control flow and exception re-raising to preserve tracebacks.
  * Minor test and fixture cleanups for consistency.

* **Bug Fixes**
  * Validation now raises more specific error types in several edge cases; tests updated accordingly.

[Review Change Stack](https://app.coderabbit.ai/change-stack/app-sre/qontract-reconcile/pull/5536)

### CODE REVIEW

**Review by esron — APPROVED** [review]
nice, LGTM

File: pyproject.toml [hunk 1]
```diff
@@ -235,8 +236,23 @@ ignore = [
     # Review later
     "FURB101", # Replace `open` with `pathlib.Path`
     "FURB103", # Replace `open` with `pathlib.Path`
+    "TRY002",  # Create your own exception class instead of raising `Exception`
+    "BLE001",  # Do not catch blind exception: `Exception`
+    "LOG015",  # call on root logger
```
[the remark this answers is not in the document]
Comment (chassing) [reply pyproject.toml:240]: This would be a huge change. postponed to the future

File: reconcile/utils/ocm/__init__.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from reconcile.utils.ocm.ocm import *  # noqa: F403
+from reconcile.utils.ocm.ocm import *
```
[the remark this answers is not in the document]
Comment (chassing) [reply reconcile/utils/ocm/__init__.py:1]: This comment is completely wrong. `ruff` automatically removed this `noqa` and doesn't complain about `F403`

**Review by hemslo — APPROVED** [review]
LGTM! Clean ruff upgrade with good `differ` consolidation.

Minor nits left as inline comments -- nothing blocking.

File: reconcile/test/external_resources/test_rds_factory.py [hunk 3]
```diff
@@ -111,7 +111,7 @@ def test_validate_timeouts_nok(timeouts: Any, expected_value_error: str) -> None
 
     module_conf = ExternalResourceModuleConfiguration(reconcile_timeout_minutes=120)
     with pytest.raises(
-        ValueError,
+        (ValueError, TypeError),
```
Comment (hemslo) [inline reconcile/test/external_resources/test_rds_factory.py:114]: nit: Since the production code now raises `TypeError`, this could assert just `TypeError` instead of accepting both.

File: pyproject.toml [hunk 4]
```diff
@@ -212,6 +212,7 @@ extend-select = [
     "DTZ",
 ]
 ignore = [
+
```
Comment (hemslo) [inline pyproject.toml]: nit: Stray blank line.

File: reconcile/test/external_resources/test_rds_factory.py (hunk 3 above)
Comment (chassing) [reply reconcile/test/external_resources/test_rds_factory.py:114]: Here, it's both `VauleError` and `TypeError`, e.g., for `create: 125m`

File: pyproject.toml (hunk 4 above)
Comment (chassing) [reply pyproject.toml]: done
